node/src/client.rs

```diff
@@ -0,0 +1,396 @@
+        + ProvideRuntimeApi<Block>
+        + HeaderBackend<Block>
+        + Sized
+        + Send
+        + Sync
+        + CallApiAt<Block, StateBackend = Backend::State>,
+    Client::Api: RuntimeApiCollection<StateBackend = Backend::State>,
+{
+}
+
+/// Execute something with the client instance.
+///
+/// As there exist multiple chains inside MainRT, like MainRT itself and testing runtime,
+/// there can exist different kinds of client types. As these client types differ in the generics
+/// that are being used, we can not easily return them from a function. For returning them from a
+/// function there exists [`Client`]. However, the problem on how to use this client instance still
+/// exists. This trait "solves" it in a dirty way. It requires a type to implement this trait and
+/// than the [`execute_with_client`](ExecuteWithClient::execute_with_client) function can be called
+/// with any possible client instance.
+///
+/// In a perfect world, we could make a closure work in this way.
+pub trait ExecuteWithClient {
+    /// The return type when calling this instance.
+    type Output;
+
+    /// Execute whatever should be executed with the given client instance.
+    fn execute_with_client<Client, Api, Backend>(self, client: Arc<Client>) -> Self::Output
+    where
+        <Api as sp_api::ApiExt<Block>>::StateBackend: sp_api::StateBackend<BlakeTwo256>,
+        Backend: sc_client_api::Backend<Block> + 'static,
+        Backend::State: sp_api::StateBackend<BlakeTwo256>,
+        Api: RuntimeApiCollection<StateBackend = Backend::State>,
+        Client: AbstractClient<Block, Backend, Api = Api> + 'static;
+}
+
+/// A handle to a MainRT client instance.
+///
+/// The MainRT service supports multiple different runtimes (MainRT itself or testing runtime). As each runtime has a
+/// specialized client, we need to hide them behind a trait. This is this trait.
+///
+/// When wanting to work with the inner client, you need to use `execute_with`.
+///
+/// See [`ExecuteWithClient`](trait.ExecuteWithClient.html) for more information.
+pub trait ClientHandle {
+    /// Execute the given something with the client.
+    fn execute_with<T: ExecuteWithClient>(&self, t: T) -> T::Output;
+}
+
+/// A client instance of MainRT.
+///
+/// See [`ExecuteWithClient`] for more information.
+#[derive(Clone)]
+pub enum Client {
+    MainRT(Arc<FullClient<main_runtime::RuntimeApi, MainExecutorDispatch>>),
+    StakingRT(Arc<FullClient<staking_runtime::RuntimeApi, StakingExecutorDispatch>>),
+}
+
+impl ClientHandle for Client {
+    fn execute_with<T: ExecuteWithClient>(&self, t: T) -> T::Output {
+        match self {
+            Self::MainRT(client) => T::execute_with_client::<_, _, FullBackend>(t, client.clone()),
+            Self::StakingRT(client) => {
+                T::execute_with_client::<_, _, FullBackend>(t, client.clone())
+            }
+        }
+    }
+}
+
+impl sc_client_api::UsageProvider<Block> for Client {
```

Comment:
If we only take the FullClient out of a Client entry and pass it to the places where they need a client, it would then already have these traits implemented for. So we should be able to avoid implementing the traits like this for the enum. What do you think @rajesh-nodle?

Reply:
Good Question, :-). Right now client is adopted to upstream code.